did/ion/request.go

```diff
@@ -0,0 +1,302 @@
+		return nil, err
+	}
+
+	toBeSigned := struct {
+		RecoveryCommitment string              `json:"recoveryCommitment"`
+		RecoveryKey        crypto.PublicKeyJWK `json:"recoveryKey"`
+		DeltaHash          string              `json:"deltaHash"`
+	}{
+		RecoveryCommitment: recoveryCommitment,
+		RecoveryKey:        recoveryKey,
+		DeltaHash:          deltaHash,
+	}
+	signedJWT, err := signer.SignJWT(toBeSigned)
+	if err != nil {
+		return nil, err
+	}
+	return &RecoverRequest{
+		Type:        Recover,
+		DIDSuffix:   didSuffix,
+		RevealValue: revealValue,
+		Delta:       delta,
+		SignedData:  signedJWT,
+	}, nil
+}
+
+type StateChange struct {
+	ServicesToAdd        []Service
+	ServiceIDsToRemove   []string
+	PublicKeysToAdd      []PublicKey
+	PublicKeyIDsToRemove []string
+}
+
+func (s StateChange) IsValid() error {
+	// check if services are valid
+	// build index of services to make sure IDs are unique
+	services := make(map[string]Service)
+	for _, service := range s.ServicesToAdd {
+		if _, ok := services[service.ID]; ok {
+			return fmt.Errorf("service %s duplicated", service.ID)
+		}
+
+		if len(service.ID) > maxIDLength {
+			return fmt.Errorf("service<%s> id is too long", service.ID)
+		}
+
+		// make sure service is valid if it's not a dupe
+		if len(service.Type) > maxServiceTypeLength {
+			return fmt.Errorf("service<%s> type %s is too long", service.ID, service.Type)
+		}
+
+		services[service.ID] = service
+	}
+
+	// check if public keys are valid
+	// build index of public keys to add
+	publicKeys := make(map[string]PublicKey)
+	for _, publicKey := range s.PublicKeysToAdd {
+		if _, ok := publicKeys[publicKey.ID]; ok {
+			return fmt.Errorf("public key<%s> is duplicated", publicKey.ID)
+		}
+
+		if len(publicKey.ID) > maxIDLength {
+			return fmt.Errorf("public key<%s> id is too long", publicKey.ID)
+		}
+
+		publicKeys[publicKey.ID] = publicKey
+	}
+
+	// check if services to remove are valid
+	for _, serviceID := range s.ServiceIDsToRemove {
+		if _, ok := services[serviceID]; ok {
+			return fmt.Errorf("service<%s> added and removed in same request", serviceID)
+		}
+
+		if len(serviceID) > maxIDLength {
+			return fmt.Errorf("service<%s> id is too long", serviceID)
+		}
+	}
+
+	// check if public keys to remove are valid
+	for _, publicKeyID := range s.PublicKeyIDsToRemove {
+		if _, ok := publicKeys[publicKeyID]; ok {
+			return fmt.Errorf("public key<%s> added and removed in same request", publicKeyID)
+		}
+
+		if len(publicKeyID) > maxIDLength {
+			return fmt.Errorf("public key<%s> id is too long", publicKeyID)
+		}
+	}
+	return nil
+}
+
+// NewUpdateRequest creates a new update request https://identity.foundation/sidetree/spec/#update
+func NewUpdateRequest(didSuffix string, updateKey, nextUpdateKey crypto.PublicKeyJWK, signer BTCSignerVerifier, stateChange StateChange) (*UpdateRequest, error) {
+	if err := stateChange.IsValid(); err != nil {
+		return nil, err
+	}
+
+	// construct update patches
+	var patches []any
```

Comment:
Hint the container capacity
```suggestion
	patches := make([]any, 0, len(stateChange.ServicesToAdd) + ...)
```